File: server/db/ingest.ts
// Loads data/seed.json into Postgres: applies the schema, then normalises the
// mocked vendor payloads into repositories, append-only scans, per-finding rows,
// component inventory and a risk snapshot.
//
// Two deliberate corrections happen here, because the relational model will not
// tolerate the shortcuts the flat file allowed:
//
//  * `lastScanDaysAgo` (a relative integer that rots) becomes an absolute
//    `captured_at`, and the 12-week trend is anchored to the last scan rather
//    than to "today" — so a repo with a stale scan truthfully has a trend that
//    stops early instead of a suspiciously fresh one.
//  * Per-severity counts with a single `oldestDays` become individual finding
//    rows whose ages span up to that oldest value, which is what makes SLA
//    ageing and MTTR real rather than approximated.
import { readFileSync } from 'node:fs';
import { join } from 'node:path';
import type { Portfolio, Repo, Severity } from '../../shared/types.ts';
import { pool, query } from './connect.ts';

const DAY = 86_400_000;
const SEVERITIES: Severity[] = ['critical', 'serious', 'moderate', 'low'];
const FORMULA_VERSION = 'v2-asymptotic';

/** Deterministic per-repo jitter, so ingest is reproducible without a seeded PRNG. */
function hash(text: string): number {
  let h = 2166136261;
  for (let i = 0; i < text.length; i += 1) {
    h = Math.imul(h ^ text.charCodeAt(i), 16777619);
  }
  return (h >>> 0) / 4294967296;
}

const VULNERABLE = [
  'org.apache.logging.log4j:log4j-core', 'com.fasterxml.jackson.core:jackson-databind',
  'org.springframework:spring-web', 'org.yaml:snakeyaml', 'commons-collections:commons-collections',
  'org.apache.tomcat.embed:tomcat-embed-core', 'ch.qos.logback:logback-classic',
  'org.bouncycastle:bcprov-jdk15on', 'io.netty:netty-codec', 'org.apache.commons:commons-text',
];

function iso(base: number, daysAgo: number): string {
  return new Date(base - daysAgo * DAY).toISOString();
}

async function ingestRepo(repo: Repo, generatedAt: number): Promise<void> {
  const [{ id }] = await query<{ id: string }>(
    `insert into repository (name, team, domain, tier, loc)
     values ($1, $2, $3, $4, $5)
     on conflict (name) do update set team = excluded.team
     returning id`,
    [repo.name, repo.team, repo.domain, repo.tier, repo.loc],
  );

  const sonarLast = repo.sonar.lastScanDaysAgo;

  // Weekly SonarQube history, newest first, anchored to the last scan date.
  const trend = [...repo.history].reverse();
  for (const [weeksBack, point] of trend.entries()) {
    const daysAgo = sonarLast + weeksBack * 7;
    await query(
      `insert into scan (repository_id, tool, captured_at, payload, gate, coverage,
                         new_code_coverage, tech_debt_days)
       values ($1, 'sonarqube', $2, $3, $4, $5, $6, $7)
       on conflict do nothing`,
      [
        id, iso(generatedAt, daysAgo),
        JSON.stringify(weeksBack === 0 ? repo.sonar : { ...repo.sonar, coverage: point.coverage }),
        weeksBack === 0 ? repo.sonar.gate : null,
        point.coverage,
        weeksBack === 0 ? repo.sonar.newCodeCoverage : null,
        weeksBack === 0 ? repo.sonar.techDebtDays : null,
      ],
    );
  }

  await query(
    `insert into scan (repository_id, tool, captured_at, payload) values ($1, 'checkmarx', $2, $3)`,
    [id, iso(generatedAt, repo.checkmarx.lastScanDaysAgo), JSON.stringify(repo.checkmarx)],
  );
  await query(
    `insert into scan (repository_id, tool, captured_at, payload, component_count, licence_violations)
     values ($1, 'sonatype', $2, $3, $4, $5)`,
    [id, iso(generatedAt, repo.sonatype.lastScanDaysAgo), JSON.stringify(repo.sonatype),
     repo.sonatype.components, repo.sonatype.licenceViolations],
  );
  await query(
    `insert into scan (repository_id, tool, captured_at, payload, spring_boot,
                       spring_boot_support, java_version, java_lts, has_codeowners)
     values ($1, 'repo-analysis', $2, $3, $4, $5, $6, $7, $8)`,
    [id, iso(generatedAt, 0), JSON.stringify(repo.analysis), repo.analysis.springBoot,
     repo.analysis.springBootSupport, repo.analysis.java, repo.analysis.javaLts,
     repo.analysis.hasCodeowners],
  );

  // Expand aggregate counts into individual findings. The oldest finding in a
  // severity sits exactly at that severity's reported age; the rest spread newer.
  let waiversLeft = repo.sonatype.waived;
  for (const [tool, counts, ages, categories] of [
    ['checkmarx', repo.checkmarx.findings, repo.checkmarx.oldestDays,
     repo.checkmarx.topCategories.map((c) => c.name)],
    ['sonatype', repo.sonatype.violations, repo.sonatype.oldestDays, ['security-policy', 'licence-policy']],
  ] as const) {
    for (const severity of SEVERITIES) {
      const count = counts[severity];
      const oldest = ages[severity];
      for (let i = 0; i < count; i += 1) {
        const daysAgo = count === 1 ? oldest : Math.round(oldest * (1 - i / count));
        // Waivers land on the least severe findings first, as they do in practice.
        const waived = waiversLeft > 0 && (severity === 'low' || severity === 'moderate');
        if (waived) waiversLeft -= 1;
        await query(
          `insert into finding (repository_id, tool, severity, category, first_seen_at, waived)
           values ($1, $2, $3, $4, $5, $6)`,
          [id, tool, severity, categories[i % categories.length], iso(generatedAt, daysAgo), waived],
        );
      }
    }
  }

  // Component inventory for the vulnerable set — enough to answer the incident
  // question ("who ships this component") across the estate.
  const offset = Math.floor(hash(repo.name) * VULNERABLE.length);
  const used = 3 + Math.floor(hash(`${repo.name}:n`) * 4);
  for (let i = 0; i < used; i += 1) {
    const coordinates = VULNERABLE[(offset + i) % VULNERABLE.length];
    const major = 1 + Math.floor(hash(`${repo.name}${coordinates}`) * 3);
    const minor = Math.floor(hash(`${coordinates}${repo.name}`) * 20);
    const isWorst = repo.sonatype.worstCve?.component === coordinates;
    await query(
      `insert into component_usage (repository_id, coordinates, version, direct, cve_id, cvss)
       values ($1, $2, $3, $4, $5, $6) on conflict (repository_id, coordinates) do nothing`,
      [id, coordinates, `${major}.${minor}.${Math.floor(hash(coordinates) * 9)}`,
       i < 2, isWorst ? repo.sonatype.worstCve?.id : null,
       isWorst ? repo.sonatype.worstCve?.cvss : null],
    );
  }
  // Ensure the reported worst CVE is always present, even if not in the sampled set.
  if (repo.sonatype.worstCve) {
    const { component, id: cve, cvss } = repo.sonatype.worstCve;
    await query(
      `insert into component_usage (repository_id, coordinates, version, direct, cve_id, cvss)
       values ($1, $2, $3, true, $4, $5)
       on conflict (repository_id, coordinates)
       do update set cve_id = excluded.cve_id, cvss = excluded.cvss`,
      [id, component, `${1 + Math.floor(hash(cve) * 3)}.${Math.floor(hash(component) * 20)}.0`, cve, cvss],
    );
  }

  // Findings that were open at the start of the trend window but are closed now.
  // Without these every critical looks like it is still open, so the burn-down can
  // only rise and MTTR is uncomputable.
  const openNow = repo.checkmarx.findings.critical + repo.sonatype.violations.critical;
  const resolvedInWindow = Math.max(0, repo.history[0].criticals - openNow);
  for (let i = 0; i < resolvedInWindow; i += 1) {
    const seenDaysAgo = 84 + Math.round(hash(`${repo.name}seen${i}`) * 120);
    // Spread closures across the 12-week window so the weekly open-count declines.
    const resolvedDaysAgo = Math.round(84 * (1 - (i + 1) / (resolvedInWindow + 1)));
    await query(
      `insert into finding (repository_id, tool, severity, category, first_seen_at, resolved_at)
       values ($1, 'sonatype', 'critical', 'security-policy', $2, $3)`,
      [id, iso(generatedAt, seenDaysAgo), iso(generatedAt, resolvedDaysAgo)],
    );
  }

  await query(
    `insert into risk_snapshot (repository_id, computed_at, score, band, sla_breaches, drivers, formula_version)
     values ($1, $2, $3, $4, $5, $6, $7)`,
    [id, iso(generatedAt, 0), repo.risk.score, repo.risk.band, repo.risk.slaBreaches,
     JSON.stringify(repo.risk.drivers), FORMULA_VERSION],
  );
}

export async function ingest(): Promise<number> {
  const schema = readFileSync(join(import.meta.dirname, 'schema.sql'), 'utf8');
  await pool.query(schema);

  const portfolio = JSON.parse(
    readFileSync(join(import.meta.dirname, '..', '..', 'data', 'seed.json'), 'utf8'),
  ) as Portfolio;
  const generatedAt = new Date(portfolio.generatedAt).getTime();

  for (const repo of portfolio.repos) await ingestRepo(repo, generatedAt);
  return portfolio.repos.length;
}

if (process.argv[1]?.endsWith('ingest.ts')) {
  const n = await ingest();
  const [counts] = await query<Record<string, string>>(
    `select (select count(*) from repository)::text as repositories,
            (select count(*) from scan)::text       as scans,
            (select count(*) from finding)::text    as findings,
            (select count(*) from component_usage)::text as components,
            (select count(*) from risk_snapshot)::text   as risk_snapshots`,
  );
  console.log(`ingested ${n} repositories`, counts);
  await pool.end();
}

File: backend/ingest.py
"""Loads data/seed.json into Postgres.

Applies the schema, then normalises the mocked vendor payloads into repositories,
append-only scans, per-finding rows, component inventory and a risk snapshot.

Risk is computed HERE (risk.py) rather than read from the seed file: the seed holds
raw vendor metrics, which is what a real SonarQube or Checkmarx API returns.

Four corrections the relational model forces, carried over from the TypeScript
version they were first found in:

  * `lastScanDaysAgo` (a relative integer that rots) becomes an absolute
    `captured_at`, and the weekly trend is anchored to the last scan rather than to
    "today" — so a repo with a stale scan truthfully has a trend that stops early.
  * Per-severity counts with a single `oldestDays` become individual finding rows,
    which is what makes SLA ageing and MTTR real rather than approximated.
  * Findings open at the start of the window but closed now get a `resolved_at`,
    so the critical burn-down can fall instead of only ever rising.
  * One row per artifact per repository, so a CVE cannot attach to two versions.
"""

import json
import sys
from datetime import datetime, timedelta
from pathlib import Path
from typing import Any

from db import pool, query
from models import SLA_DAYS
from risk import FORMULA_VERSION, compute_risk

ROOT = Path(__file__).resolve().parent.parent
SEED = ROOT / "data" / "seed.json"
SCHEMA = Path(__file__).resolve().parent / "schema.sql"

SEVERITIES = ("critical", "serious", "moderate", "low")

VULNERABLE = [
    "org.apache.logging.log4j:log4j-core",
    "com.fasterxml.jackson.core:jackson-databind",
    "org.springframework:spring-web",
    "org.yaml:snakeyaml",
    "commons-collections:commons-collections",
    "org.apache.tomcat.embed:tomcat-embed-core",
    "ch.qos.logback:logback-classic",
    "org.bouncycastle:bcprov-jdk15on",
    "io.netty:netty-codec",
    "org.apache.commons:commons-text",
]


def _hash(text: str) -> float:
    """Deterministic per-repo jitter, so ingest is reproducible without a PRNG."""
    h = 2166136261
    for ch in text:
        h = ((h ^ ord(ch)) * 16777619) & 0xFFFFFFFF
    return h / 0x100000000


def _at(base: datetime, days_ago: float) -> datetime:
    return base - timedelta(days=days_ago)


def _ingest_repo(cur: Any, repo: dict[str, Any], base: datetime) -> None:
    cur.execute(
        """insert into repository (name, team, domain, tier, loc)
           values (%s, %s, %s, %s, %s)
           on conflict (name) do update set team = excluded.team
           returning id""",
        (repo["name"], repo["team"], repo["domain"], repo["tier"], repo["loc"]),
    )
    repo_id = cur.fetchone()["id"]

    sonar, cx, sc, analysis = (
        repo["sonar"],
        repo["checkmarx"],
        repo["sonatype"],
        repo["analysis"],
    )
    sonar_last = sonar["lastScanDaysAgo"]

    # Weekly SonarQube history, newest first, anchored to the last scan date.
    for weeks_back, point in enumerate(reversed(repo["history"])):
        payload = sonar if weeks_back == 0 else {**sonar, "coverage": point["coverage"]}
        cur.execute(
            """insert into scan (repository_id, tool, captured_at, payload, gate,
                                 coverage, new_code_coverage, tech_debt_days)
               values (%s, 'sonarqube', %s, %s, %s, %s, %s, %s)
               on conflict do nothing""",
            (
                repo_id,
                _at(base, sonar_last + weeks_back * 7),
                json.dumps(payload),
                sonar["gate"] if weeks_back == 0 else None,
                point["coverage"],
                sonar["newCodeCoverage"] if weeks_back == 0 else None,
                sonar["techDebtDays"] if weeks_back == 0 else None,
            ),
        )

    cur.execute(
        """insert into scan (repository_id, tool, captured_at, payload)
           values (%s, 'checkmarx', %s, %s)""",
        (repo_id, _at(base, cx["lastScanDaysAgo"]), json.dumps(cx)),
    )
    cur.execute(
        """insert into scan (repository_id, tool, captured_at, payload,
                             component_count, licence_violations)
           values (%s, 'sonatype', %s, %s, %s, %s)""",
        (
            repo_id,
            _at(base, sc["lastScanDaysAgo"]),
            json.dumps(sc),
            sc["components"],
            sc["licenceViolations"],
        ),
    )
    cur.execute(
        """insert into scan (repository_id, tool, captured_at, payload, spring_boot,
                             spring_boot_support, java_version, java_lts, has_codeowners)
           values (%s, 'repo-analysis', %s, %s, %s, %s, %s, %s, %s)""",
        (
            repo_id,
            base,
            json.dumps(analysis),
            analysis["springBoot"],
            analysis["springBootSupport"],
            analysis["java"],
            analysis["javaLts"],
            analysis["hasCodeowners"],
        ),
    )

    # Expand aggregate counts into individual findings. The oldest finding in a
    # severity sits exactly at that severity's reported age; the rest spread newer.
    waivers_left = sc["waived"]
    sources = (
        ("checkmarx", cx["findings"], cx["oldestDays"],
         [c["name"] for c in cx["topCategories"]] or ["uncategorised"]),
        ("sonatype", sc["violations"], sc["oldestDays"],
         ["security-policy", "licence-policy"]),
    )
    for tool, counts, ages, categories in sources:
        for severity in SEVERITIES:
            count, oldest = counts[severity], ages[severity]
            for i in range(count):
                days_ago = oldest if count == 1 else round(oldest * (1 - i / count))
                # Waivers land on the least severe findings first, as in practice.
                waived = waivers_left > 0 and severity in ("low", "moderate")
                if waived:
                    waivers_left -= 1
                cur.execute(
                    """insert into finding (repository_id, tool, severity, category,
                                            first_seen_at, waived)
                       values (%s, %s, %s, %s, %s, %s)""",
                    (repo_id, tool, severity, categories[i % len(categories)],
                     _at(base, days_ago), waived),
                )

    # Findings open at the start of the trend window but closed now, so the
    # burn-down can fall and MTTR is computable.
    open_now = cx["findings"]["critical"] + sc["violations"]["critical"]
    resolved_in_window = max(0, repo["history"][0]["criticals"] - open_now)
    for i in range(resolved_in_window):
        seen = 84 + round(_hash(f"{repo['name']}seen{i}") * 120)
        closed = round(84 * (1 - (i + 1) / (resolved_in_window + 1)))
        cur.execute(
            """insert into finding (repository_id, tool, severity, category,
                                    first_seen_at, resolved_at)
               values (%s, 'sonatype', 'critical', 'security-policy', %s, %s)""",
            (repo_id, _at(base, seen), _at(base, closed)),
        )

    # Component inventory for the vulnerable set — enough to answer the incident
    # question ("who ships this component") across the estate.
    offset = int(_hash(repo["name"]) * len(VULNERABLE))
    used = 3 + int(_hash(f"{repo['name']}:n") * 4)
    worst = sc["worstCve"]
    for i in range(used):
        coordinates = VULNERABLE[(offset + i) % len(VULNERABLE)]
        major = 1 + int(_hash(f"{repo['name']}{coordinates}") * 3)
        minor = int(_hash(f"{coordinates}{repo['name']}") * 20)
        is_worst = bool(worst and worst["component"] == coordinates)
        cur.execute(
            """insert into component_usage (repository_id, coordinates, version,
                                            direct, cve_id, cvss)
               values (%s, %s, %s, %s, %s, %s)
               on conflict (repository_id, coordinates) do nothing""",
            (repo_id, coordinates, f"{major}.{minor}.{int(_hash(coordinates) * 9)}",
             i < 2, worst["id"] if is_worst else None,
             worst["cvss"] if is_worst else None),
        )
    # Ensure the reported worst CVE is present, attached to the existing row.
    if worst:
        cur.execute(
            """insert into component_usage (repository_id, coordinates, version,
                                            direct, cve_id, cvss)
               values (%s, %s, %s, true, %s, %s)
               on conflict (repository_id, coordinates)
               do update set cve_id = excluded.cve_id, cvss = excluded.cvss""",
            (repo_id, worst["component"],
             f"{1 + int(_hash(worst['id']) * 3)}.{int(_hash(worst['component']) * 20)}.0",
             worst["id"], worst["cvss"]),
        )

    # The score AS REPORTED at ingest time, with the formula version that produced it.
    scored = compute_risk(repo)
    cur.execute(
        """insert into risk_snapshot (repository_id, computed_at, score, band,
                                      sla_breaches, drivers, formula_version)
           values (%s, %s, %s, %s, %s, %s, %s)""",
        (repo_id, base, scored["score"], scored["band"], scored["slaBreaches"],
         json.dumps(scored["drivers"]), FORMULA_VERSION),
    )


def ingest() -> int:
    seed = json.loads(SEED.read_text())
    base = datetime.fromisoformat(seed["generatedAt"].replace("Z", "+00:00"))

    with pool.connection() as conn:
        # Idempotent DDL, then clear the data. Kept separate because dropping and
        # recreating the enum types changed their OIDs, which broke pooled psycopg
        # connections holding the old ones ("cache lookup failed for type NNNNN").
        conn.execute(SCHEMA.read_text())
        conn.execute(
            "truncate risk_snapshot, component_usage, finding, scan, repository"
            " restart identity cascade"
        )

    from psycopg.rows import dict_row

    with pool.connection() as conn, conn.cursor(row_factory=dict_row) as cur:
        for repo in seed["repos"]:
            _ingest_repo(cur, repo, base)

    return len(seed["repos"])


if __name__ == "__main__":
    pool.open()
    n = ingest()
    counts = query(
        """select (select count(*) from repository) as repositories,
                  (select count(*) from scan) as scans,
                  (select count(*) from finding) as findings,
                  (select count(*) from component_usage) as components,
                  (select count(*) from risk_snapshot) as risk_snapshots"""
    )[0]
    print(f"ingested {n} repositories", counts, file=sys.stderr)
    pool.close()
